GAWB-1173 Calculate Google Bucket Usage

breilly2 opened the pull request:
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  - Acceptance criteria exists and is met
  - Note any changes to implementation from the description
  - Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  - If PR includes new or changed db queries, include the explain plans in the description
  - Make sure liquibase is updated if appropriate
  - If doing a migration, take a backup of the
    [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
    and
    [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
    DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  - Authentication
  - Authorization
  - Encryption
  - Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
- Review cycle:
  - LR reviews
  - Rest of team may comment on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH
  - Submitter updates documentation as needed
  - Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

breilly2 committed:
Added GoogleServicesDAO method for calculating storage usage for a bucket

breilly2 committed:
Added Workspace service for calculating storage usage for a bucket

breilly2 commented:
Please comment on anything, including suggestions to better follow Scala idioms. I'm learning here, so not mentioning an improvement is going to mean I keep doing things the wrong way. Thanks.

MatthewBemis commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
+        result match {
+          case r if r.getItems == null => BigInt(0)
+          case r if r.getNextPageToken == null => sumItemSizes(r.getItems)
+          case r => sumItemSizes(r.getItems) + Await.result(processPage(Some(result.getNextPageToken)), Duration.Inf)
```
Await.result is a blocking task so generally we only use it in tests and in production code that is very rarely used by admins. this endpoint could be called every time a user loads a workspace in the UI so we want to avoid that here

breilly2 replied:
I thought I thought this through and decided it was okay here, but maybe I missed something. Yes, I am blocking on the result for the next page, but that's somewhat necessary in order to do the addition. Also, retryWithRecoverWhen500orGoogleError() is already returning a Future, so the only thing being blocked here fulfilling that Future which necessarily needs to wait.

I thought about finding a way to get rid of the Await.result(), and would be happy to for cleanliness. Still new to this, so suggestions are welcome.

geoffjentry replied:
In my neck of the woods, nearly every single time someone has convinced me that `Await.result` is fine in the context it was in has led to a horrible problem. I still end up getting convinced it is fine. I'm not smart.

dvoet replied:
Unfortunately, it is not fine here either. The problem is that Await blocks the thread which can lead to resource contention/starvation.

The solution is to chain futures together with map or flatMap. I would try something like

```
def processPage(sum: BigInt, pageToken: Option[String] = None): Future[(BigInt, Option[String])] = {
  val fetcher = getStorage(getBucketServiceAccountCredential).objects().list(bucketName)
  maxResults.forEach { fetcher setMaxResults maxResults.get }
  pageToken.forEach { fetcher setPageToken pageToken.get }

  retryWhen500orGoogleError(() => {
    val result = executeGoogleRequest(fetcher)
    (Option(result.getItems).map(sumItemSizes(_.getItems)).getOrElse(BigInt(0)), Option(result.getNextPageToken))
  }) flatMap { 
    case (pageSize, None) => Future.successful(sum + pageSize)
    case (pageSize, Some(nextPageToken)) => processPage(sum + pageSize, Option(nextPageToken))
  }
}
```

breilly2 commented on src/main/resources/swagger/rawls.yaml:
```diff
@@ -3247,6 +3247,35 @@ paths:
             - email
             - profile
             - https://www.googleapis.com/auth/devstorage.full_control
+  '/api/workspaces/{workspaceNamespace}/{workspaceName}/getBucketUsage':
```
It just occurred to me that "getBucketUsage" is a horrible name because "get" is redundant with the HTTP method. I'll change this to "bucketUsage" unless someone can suggest a better name. Since this is an exported API, others who have been around longer probably have stronger feelings about this.

geoffjentry commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala:
```diff
@@ -62,6 +62,20 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
+    * @param maxResults (optional) the page size to use when fetching objects
+    * @return the size in bytes of the data stored in the bucket
+    */
+  def getBucketUsage(bucketName: String, maxResults: Option[Long] = None): Future[BigInt]
```
I'm only butting in because I saw you ask about scala idioms, anything I say here might contradict Rawls idioms. I also made no effort to look at surrounding code, potentially compounding the Rawls idiom effect. YMMV.

Generally `get` is frowned upon in scala, it's more typical to just have `bucketUsage` in a case like this.

breilly2 replied:
I'm not crazy about the "get" prefix here, but I think I'll leave it for now since most other functions in `GoogleServicesDAO` have it. I also don't want to imply that it's a simple property access. Maybe just being in `GoogleServicesDAO` is enough, or maybe we could lean toward a different prefix verb such as "fetch".

geoffjentry replied:
Like I said, I have no idea on rawls idioms :)

geoffjentry commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
 
+  override def getBucketUsage(bucketName: String, maxResults: Option[Long]): Future[BigInt] = {
+    def sumItemSizes(items: java.util.List[StorageObject]) = {
+      (BigInt(0) /: items) { (sum, o) => sum + o.getSize }
```
IMO leave the operator characters in the past as a bad memory of days gone by in scala, stick w/ `fold`

breilly2 replied:
Thanks. In this case, I think I need to `foldLeft` or `foldRight`:

```
items.foldLeft(BigInt(0)) { (sum, o) => sum + o.getSize }
```

The reason is that `items` is an array of some object type rather than an array of `BigInt`. Is that okay or should I first `map` it to an array of `BigInt` and then `fold` it?

dvoet replied:
agreed, what is /: anyway. You can probably use reduce instead
`items.map(_.getSize).reduce(_ + _)`

breilly2 replied:
According to the book I was looking at, `/:` is a `foldLeft` operator.

There's some extra complexity that `getSize` returns a `java.math.BigInteger`. By being a little more verbose and aggressively moving to `BigInt`s as soon as possible, `reduce(_ + _)` can actually become `sum`.

```
items.map((s: StorageObject) => BigInt(s.getSize)).sum
```

geoffjentry commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
+
+    def processPage(pageToken: Option[String] = None): Future[BigInt] = {
+      val fetcher = getStorage(getBucketServiceAccountCredential).objects().list(bucketName)
+      if (maxResults.isDefined) fetcher setMaxResults maxResults.get
```
I'm not 100% what these 2 lines are doing but I think you can roll with a `foreach` 

If you haven't seen it, http://blog.originate.com/blog/2014/06/15/idiomatic-scala-your-options-do-not-match/ has a phenomenal cheatsheet for options

geoffjentry commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
+      }
+    }
+
+    processPage()
```
The `()` is typically used to signify side effects which I don't think is the case here

breilly2 replied:
I was getting an error without the () because of the one optional parameter. I agree, I wish I could remove the (). I'm also not entirely satisfied with the code. I could remove the default parameter value and explicitly pass in None here, but then I don't like how that reads. I may see if I can make it more clear by splitting out fetchFirstPage() and fetchNextPage(String) functions.

geoffjentry commented on src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -228,6 +228,8 @@ case class WorkspacePermissionsPair(workspaceId: String,
 
 case class WorkspaceStatus(workspaceName: WorkspaceName, statuses: Map[String, String])
 
+case class BucketUsageResponse(usageInBytes: BigInt)
```
case classes should always be marked as final

breilly2 replied:
@MatthewBemis do you think there would be objection to adding `final` to all of these? It's either that or leave it as is. I'm not going to break the established practice by making only this one final.

MatthewBemis replied:
i personally don't object to this but i'll defer to @dvoet 

we have a bunch of model classes so we'd want to change them there too

dvoet replied:
meh, none of our other case classes are final... just don't extend them and we will be fine

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala:
```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
+      if (maxResults.isDefined) fetcher setMaxResults maxResults.get
+      if (pageToken.isDefined) fetcher setPageToken pageToken.get
+
+      retryWithRecoverWhen500orGoogleError(() => {
```
you don't have any recovery code so you can just call retryWhen500orGoogleError and remove the error handling block at the end

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1466,6 +1468,16 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    dataSource.inTransaction { dataAccess =>
+      withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
+        DBIO.from(gcsDAO.getBucketUsage(workspaceContext.workspace.bucketName)) map { usage =>
+          RequestComplete(BucketUsageResponse(usage))
```
we have learned that it is a bad idea to include external service calls in a transaction. Try

```
def getBucketUsage(workspaceName: WorkspaceName): Future[PerRequestMessage] = {
  dataSource.inTransaction { dataAccess =>
    withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
      workspaceContext.workspace.bucketName
    }
  } flatMap { bucketName =>
    gcsDAO.getBucketUsage(bucketName))
  } map { usage =>
    RequestComplete(BucketUsageResponse(usage))
  }
}
```

or better yet

```
def getBucketUsage(workspaceName: WorkspaceName): Future[PerRequestMessage] = {
  for {
    bucketName <- dataSource.inTransaction { dataAccess =>
      withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
        workspaceContext.workspace.bucketName
      }
    }

    usage <- gcsDAO.getBucketUsage(bucketName))

  } yield {
    RequestComplete(BucketUsageResponse(usage))
  }
}
```

breilly2 replied:
I knew this would come up but wasn't sure how to handle it. Thanks for the suggestions.

Initially, I didn't like the "better yet"; `for` implied to me working with a collection. However, the notion of having sequential generators and an eventual yield reads very nicely... better than the `flatMap`/`map` (which actually is no better than `for` in terms of implying working with a collection).

I did have to use `DBIO.from(Future.successful(...)` to get the bucket name out of the transaction context, but that makes sense.

dvoet commented on src/test/scala/org/broadinstitute/dsde/rawls/integrationtest/HttpGoogleServicesDAOSpec.scala:
```diff
@@ -329,6 +329,30 @@ class HttpGoogleServicesDAOSpec extends FlatSpec with Matchers with IntegrationT
+  it should "calculate data size" in {
+    val random = scala.util.Random
+    val testUser = testCreator.copy(userSubjectId = random.nextLong().toString)
+    val googleWorkspaceInfo = Await.result(gcsDAO.setupWorkspace(testUser, testProject, UUID.randomUUID.toString, testWorkspace, None), Duration.Inf)
```
As I said with face-words, you don't need a workspace to test this, just a bucket with some stuff in it.

dvoet reviewed (changes requested):
also add a test to WorkspaceApiServiceSpec that everything besides the actual google call works

breilly2 committed:
Did away with Await.result when fetching multiple pages of storage objects.
Moved call to GCS outside of DB transaction.
Removed ugly "get" from the bucket usage web service method.

breilly2 committed:
Added workspace API tests for bucket usage service.
Changed bucket usage integration test to only create a bucket instead of an entire workspace.

MatthewBemis commented on src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:
```diff
@@ -1847,5 +1847,39 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
+      }
+  }
+
+  // TODO: adjust user access for this test once the requirements of GAWB-1145 settle down
```
what parts of the requirements are still fluid?

breilly2 replied:
There is some discussion about which types of users should see this information. No changes have made their way from the comment thread to the description yet, but there is talk of only showing this to owners or owners and writers.

dvoet commented on src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -1466,6 +1468,19 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+    for {
+      bucketName <- dataSource.inTransaction { dataAccess =>
+        withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Read, dataAccess) { workspaceContext =>
+          DBIO.from(Future.successful(workspaceContext.workspace.bucketName))
```
DBIO.successful instead of DBIO.from and Future.successful

breilly2 committed:
Changed DBIO.from(Future.successful... to DBIO.successful(...

breilly2 commented:
The new approach (using Google storage logs) seems to be working well and is much more viable. I'm abandoning these changes, but thank you all for your feedback.
